osu.Game.Rulesets.Mania/Skinning/LegacyBodyPiece.cs

```diff
@@ -60,28 +95,68 @@ private void load(ISkinSource skin, IScrollingInfo scrollingInfo, DrawableHitObj
 
         private void onIsHittingChanged(ValueChangedEvent<bool> isHitting)
         {
-            if (!(sprite is TextureAnimation animation))
-                return;
+            if (bodySprite is TextureAnimation bodyAnimation)
+            {
+                bodyAnimation.GotoFrame(0);
+                bodyAnimation.IsPlaying = isHitting.NewValue;
+            }
+
+            if (lightContainer != null)
+            {
+                if (isHitting.NewValue)
+                {
+                    // Clear the fade out and, more importantly, the removal.
+                    lightContainer.ClearTransforms();
+
+                    // Only add the container if the removal has taken place.
+                    if (lightContainer.Parent == null)
+                        Column.TopLevelContainer.Add(lightContainer);
 
-            animation.GotoFrame(0);
-            animation.IsPlaying = isHitting.NewValue;
+                    // The light must be seeked only after being loaded, otherwise a nullref occurs (https://github.com/ppy/osu-framework/issues/3847).
+                    if (light is TextureAnimation lightAnimation)
+                        lightAnimation.GotoFrame(0);
+
+                    lightContainer.FadeIn(80);
+                }
+                else
+                {
+                    lightContainer.FadeOut(120)
+                                  .OnComplete(d => Column.TopLevelContainer.Remove(d));
+                }
```

Comment:
Yeah, this does feel maybe not *bad*, but a little unorthodox, especially the add/remove... Not sure if the other approach would be better to be fair; would have to compare the end results most likely. I could potentially see presses/releases before the fade is done being more of a problem with the event approach since this code handles them correctly without much hassle.

Reply:
I think the biggest tell is whether it's going to have to be done in the default skin too. Though the default skin may also put it inside the hold note itself rather than at the explosion layer, so we'll have to see...